packages/substrate-client/tests/chainHead-functions.spec.ts

```diff
@@ -0,0 +1,106 @@
+import { expect, describe, test } from "vitest"
+import { setupChainHead, setupChainHeadWithSubscription } from "./fixtures"
+import { ErrorDisjoint } from "@/chainhead"
+import { ErrorRpc, RpcError } from "@/client"
+
+describe.each([
+  [
+    "unpin" as "unpin",
+    [["someHash", "someOtherHash"]] as [string[]],
+    null,
+    undefined,
+  ],
+  ["header" as "header", ["someHash"] as [string], "someHeader", "someHeader"],
+])("chainhead: %s", (name, args, result, expectedResult) => {
+  test("it sends the correct message", () => {
+    const {
+      fixtures: { getNewMessages, SUBSCRIPTION_ID },
+      ...chainHead
+    } = setupChainHeadWithSubscription()
+
+    getNewMessages()
+    chainHead[name](...(args as [any]))
+
+    expect(getNewMessages()).toMatchObject([
+      {
+        method: `chainHead_unstable_${name}`,
+        params: [SUBSCRIPTION_ID, ...args],
+      },
+    ])
+  })
+
+  test("it resolves the correct response", () => {
+    const {
+      fixtures: { sendMessage },
+      ...chainHead
+    } = setupChainHeadWithSubscription()
+
+    const promise = chainHead[name](...(args as [any]))
+
+    sendMessage({
+      id: 2,
+      result,
+    })
+
+    return expect(promise).resolves.toEqual(expectedResult)
+  })
+
+  test("it rejects the JSON-RPC Error when the request fails", () => {
+    const {
+      fixtures: { sendMessage },
+      ...chainHead
+    } = setupChainHeadWithSubscription()
+
+    const promise = chainHead[name](...(args as [any]))
+
+    const error: RpcError = {
+      code: -32700,
+      message: "Parse error",
+    }
+    sendMessage({
+      id: 2,
+      error,
+    })
+
+    return expect(promise).rejects.toEqual(new ErrorRpc(error))
+  })
+
+  test("it rejects with an `ErrorDisjoint` Error when the function is created after `unfollow`", () => {
+    const { unfollow, ...chainHead } = setupChainHead()
+
+    unfollow()
+
+    return expect(chainHead[name](...(args as [any]))).rejects.toEqual(
+      new ErrorDisjoint(),
+    )
+  })
+
+  test("it rejects an `ErrorDisjoint` Error when the follow subscription fails", async () => {
+    let { fixtures, ...chainHead } = setupChainHead()
+
+    const promise = chainHead[name](...(args as [any]))
+
+    fixtures.sendMessage({
+      id: 1,
+      error: {
+        code: -32700,
+        message: "Parse error",
+      },
+    })
+
+    await expect(promise).rejects.toEqual(new ErrorDisjoint())
+    ;({ fixtures, ...chainHead } = setupChainHead())
+
+    fixtures.sendMessage({
+      id: 1,
+      error: {
+        code: -32700,
+        message: "Parse error",
+      },
+    })
+
+    await expect(chainHead[name](...(args as [any]))).rejects.toEqual(
+      new ErrorDisjoint(),
+    )
+  })
```

Comment:
duplicated or intentional?

Reply:
mmmm, if we are talking about the same thing, then it's a bit of both... I mean, specially with tests I rather having duplication than creating premature abstractions... However, I can see how it would make a lot of sense to have a `parseError` export from the fixtures, so I exported this one:

```ts
export const parseError: IRpcError = {
  code: -32700,
  message: "Parse error",
}
```

Reply:
lines 79-91 and 92-104 do the same thing, so the confusion is why are we doing it twice

Reply:
oh, I see! yeah, I actually did that on purpose. I will add a comment. This is kinda related to [this](https://github.com/paritytech/capi/pull/65#discussion_r1309631442).

If you pay close attention, you will notice that they don't do exactly the same thing:

- lines 79-91 checks that the `DisjointError` takes place when the errored `followSubscription` message arrives **after** the user has already called `unpin` or `header`.
- lines 92-104 checks that the `DisjointError` takes place when the errored `followSubscription` message arrives **before** the user has called `unpin` or `header`.

Reply:
I [added a comment and I also tried to make it a bit more explicit](https://github.com/paritytech/capi/pull/65/commits/a7e271e226932bb8a991a70008fab0cd9c30e404). Thanks for pointing this out!

Reply:
ah yes you're right. it looked the same but it wasn't because promises are eager rather than lazy.